pso/src/main/scala/cilib/pso/PSO.scala

```diff
@@ -54,6 +65,81 @@ object PSO {
     if (b) updatePBest(p) else Step.point(p)
   }
 
+ def clamp[S,A:scalaz.Equal](x: Entity[S,A])(implicit N: spire.math.Numeric[A]) =
+    Lenses._position.modify((p: Position[A]) => {
+      val c: NonEmptyList[A] = p.pos zip p.boundary map {
+        case t if N.toDouble(t._1) < t._2.lowerValue => N.fromDouble(t._2.lowerValue)
+        case t if N.toDouble(t._1) > t._2.upperValue => N.fromDouble(t._2.upperValue)
+        case t if t._2.contains(N.toDouble(t._1)) => t._1
+      }
+      Lenses._vector[A].set(c)(p)
+    })(x)
+    
+ def initToPB[S,A:scalaz.Equal](x: Entity[S,A])(implicit N: spire.math.Numeric[A], M: HasMemory[S, A]) =
+    Lenses._position.modify((p: Position[A]) => {
+        val c: NonEmptyList[A] = p.pos zip p.boundary zip M._memory.get(x.state).pos map {x => flatten(x)} map {
+            case t if t._2.contains(N.toDouble(t._1)) => t._1
+            case t if t._2.doesNotContain(N.toDouble(t._1)) => t._3
+        }
+        Lenses._vector[A].set(c)(p)
+    })(x)
+
+  def initToGB[S,A:scalaz.Equal](x: Entity[S,A], gBest: Position[A])(implicit N: spire.math.Numeric[A]) = 
+    Lenses._position.modify((p: Position[A]) => {
+      val c: NonEmptyList[A] = p.pos zip p.boundary zip gBest.pos map {x => flatten(x)} map {
```

Comment:
The one thing bothering me with this code is that it is partial - this is actually true for all the functions that use the case to match the different options.

The poly function should probably be removed, and we can replace the double map with a single one. For example:

```scala
 val c: NonEmptyList[A] = p.pos zip p.boundary zip gBest.pos map { case ((p, b), g) =>
    if (b.contains(N.toDouble(p))) p else g
 }
```

The `case` in the above is only there to destructure the tuple and is unfortunately a requirement in this usage because the scala compiler is a bit lacking wrt expressive power.